File: projects/11/JackCompiler/src/Token.java
/**
 * @author Administrator
 * @since 2021/4/1
 */
class Token {
    TokenType tokenType;
    Keyword keyword;
    Object value;

    public Token(TokenType tokenType, Keyword keyword, Object value) {
        this.tokenType = tokenType;
        this.keyword = keyword;
        this.value = value;
    }

    public TokenType tokenType() {
        return tokenType;
    }
    public Keyword keyword() {
        return keyword;
    }
    public Character symbol() {
        return (Character)value;
    }
    public String identifier() {
        return (String)value;
    }
    public int intVal() {
        return (Integer)value;
    }
    public String stringVal() {
        return (String)value;
    }

    @Override
    public String toString() {
        return "Node" + '"' +
                "tokenType=" + tokenType +
                ", keyword=" + keyword +
                ", value=" + value +
                '"';
    }
}

File: projects/10/JackCompiler-OC/JackCompiler-OC/Token.m
//
//  Token.m
//  JackCompiler-OC
//
//  Created by janezhuang on 2022/2/3.
//

#import "Token.h"

@interface Token ()

@end

@implementation Token

- (id)initWithType:(TokenType)type value:(NSString *)value {
    self = [super init];
    if (self) {
        self.type = type;
        self.value = [value copy];
        
    }
    return self;
}

- (NSString *)keywordVal {
    return self.value;
}

- (NSString *)symbol {
    return self.value;
}

- (NSString *)identifier {
    return self.value;
}

- (SInt32)intVal {
    return [self.value intValue];
}

- (NSString *)stringVal {
    //去除双引号
    return [self.value substringWithRange:NSMakeRange(1, self.value.length - 2)];
}

- (BOOL)isSymbol:(NSString *)symbol {
    return self.type == TokenType_Symbol && [self.symbol isEqualToString:symbol];
}

- (BOOL)isKeyword:(Keyword)keyword {
    return self.type == TokenType_Keyword && self.keyword == keyword;
}
@end
